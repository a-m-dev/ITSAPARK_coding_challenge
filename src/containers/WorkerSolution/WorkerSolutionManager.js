import { useState, useEffect, useCallback } from "react";

import worker from "workerize-loader!../../../src/workers/computeNum"; // eslint-disable-line import/no-webpack-loader-syntax

const WorkerSolutionManager = () => {
  const [timeItTakes, setTimeItTakes] = useState(0);
  const [count, setCount] = useState(null);
  const [loading, setLoading] = useState(false);

  const workerInstance = worker();

  // effects
  useEffect(() => {
    workerInstance.addEventListener("message", (evt) => {
      if (evt.data?.result) {
        setCount(evt.data.result);
        setTimeItTakes(evt.data.totalTime);
        setLoading(false);
      }
    });
  }, [setCount, workerInstance, setTimeItTakes, setLoading]);

  // handlers
  const handleSubmitForm = useCallback(
    ({ startNum, stopNum }) => {
      console.log(">>> WORKER", { startNum, stopNum });

      setLoading(true);

      workerInstance.calculateTheRange({
        startNum: parseInt(startNum),
        stopNum: parseInt(stopNum),
        target: 3,
      });
    },
    [workerInstance, setLoading]
  );

  return {
    data: { count, timeItTakes, loading },
    actions: { handleSubmitForm },
  };
};

export default WorkerSolutionManager;
