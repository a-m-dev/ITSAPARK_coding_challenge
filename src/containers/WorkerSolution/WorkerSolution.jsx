import { NumsRangeForm, ResultBox, Loading } from "../../components";
import WorkerSolutionManager from "./WorkerSolutionManager";
import "./WorkerSolution.scss";

const WorkerSolution = () => {
  const {
    data,
    actions: { handleSubmitForm },
  } = WorkerSolutionManager();

  return (
    <article className="container worker-solution">
      <section className="worker-solution__container">
        <h1>Insert Your Range:</h1>
        <h4>Implemented brute-force way with web worker:</h4>
        <NumsRangeForm submitFormHandler={handleSubmitForm} />
        {data.loading ? <Loading /> : <ResultBox {...data} />}
      </section>
    </article>
  );
};

export default WorkerSolution;
