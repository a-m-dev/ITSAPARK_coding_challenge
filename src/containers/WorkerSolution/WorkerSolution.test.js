import { render, fireEvent, waitFor } from "@testing-library/react";
import WorkerSolution from "./WorkerSolution";
import worker from "workerize-loader!../../../src/workers/computeNum"; // eslint-disable-line import/no-webpack-loader-syntax

jest.mock("workerize-loader!../../../src/workers/computeNum", () => {
  return () => {
    return {
      addEventListener: jest.fn(),
      calculateTheRange: jest.fn(),
    };
  };
});

describe("<WorkerSolution />", () => {
  test("should match with snapshot", () => {
    const result = render(<WorkerSolution />);
    expect(result).toMatchSnapshot();
  });

  test("should render properly", () => {
    const { getByTestId } = render(<WorkerSolution />);

    const startNumberInput = getByTestId("startNumber");
    const stopNumberInput = getByTestId("stopNumber");
    const submitButton = getByTestId("submitButton");

    expect(startNumberInput).toBeTruthy();
    expect(stopNumberInput).toBeTruthy();
    expect(submitButton).toBeTruthy();

    expect(startNumberInput).toHaveValue("");
    expect(stopNumberInput).toHaveValue("");
    expect(submitButton).toHaveTextContent("Submit");
  });
});
