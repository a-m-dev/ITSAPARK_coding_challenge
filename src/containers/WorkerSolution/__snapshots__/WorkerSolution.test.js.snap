// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WorkerSolution /> should match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="container worker-solution"
      >
        <section
          class="worker-solution__container"
        >
          <h1>
            Insert Your Range:
          </h1>
          <h4>
            Implemented brute-force way with web worker:
          </h4>
          <section
            class="nums-range-form"
          >
            <form
              action="#"
            >
              <fieldset
                class="input-field"
              >
                <span
                  class="input-field__label"
                >
                  Start Number
                </span>
                <input
                  autocomplete="off"
                  class="input-field__input"
                  data-testid="startNumber"
                  name="startNum"
                  placeholder=""
                  type="text"
                  value=""
                />
              </fieldset>
              <fieldset
                class="input-field"
              >
                <span
                  class="input-field__label"
                >
                  Stop Number
                </span>
                <input
                  autocomplete="off"
                  class="input-field__input"
                  data-testid="stopNumber"
                  name="stopNum"
                  placeholder=""
                  type="text"
                  value=""
                />
              </fieldset>
              <button
                data-testid="submitButton"
                type="submit"
              >
                Submit
              </button>
            </form>
          </section>
          <hr />
          <section
            class="result-box"
          >
            <h1
              data-testid="resultCount"
            >
              0
            </h1>
            <span>
              Items found in given range!
            </span>
            <span>
              in 
              <strong
                data-testid="timeTook"
              >
                0
              </strong>
               sec
            </span>
          </section>
        </section>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="container worker-solution"
    >
      <section
        class="worker-solution__container"
      >
        <h1>
          Insert Your Range:
        </h1>
        <h4>
          Implemented brute-force way with web worker:
        </h4>
        <section
          class="nums-range-form"
        >
          <form
            action="#"
          >
            <fieldset
              class="input-field"
            >
              <span
                class="input-field__label"
              >
                Start Number
              </span>
              <input
                autocomplete="off"
                class="input-field__input"
                data-testid="startNumber"
                name="startNum"
                placeholder=""
                type="text"
                value=""
              />
            </fieldset>
            <fieldset
              class="input-field"
            >
              <span
                class="input-field__label"
              >
                Stop Number
              </span>
              <input
                autocomplete="off"
                class="input-field__input"
                data-testid="stopNumber"
                name="stopNum"
                placeholder=""
                type="text"
                value=""
              />
            </fieldset>
            <button
              data-testid="submitButton"
              type="submit"
            >
              Submit
            </button>
          </form>
        </section>
        <hr />
        <section
          class="result-box"
        >
          <h1
            data-testid="resultCount"
          >
            0
          </h1>
          <span>
            Items found in given range!
          </span>
          <span>
            in 
            <strong
              data-testid="timeTook"
            >
              0
            </strong>
             sec
          </span>
        </section>
      </section>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
